packages/common-utils/src/renderChartConfig.ts

```diff
@@ -737,21 +735,49 @@ type ChartConfigWithOptDateRangeEx = ChartConfigWithOptDateRange & {
   includedDataInterval?: string;
 };
 
-function renderWith(
+async function renderWith(
   chartConfig: ChartConfigWithOptDateRangeEx,
   metadata: Metadata,
-): ChSql | undefined {
+): Promise<ChSql | undefined> {
   const { with: withClauses } = chartConfig;
   if (withClauses) {
     return concatChSql(
       ',',
-      withClauses.map(clause => {
-        if (clause.isSubquery === false) {
-          return chSql`(${clause.sql}) AS ${{ Identifier: clause.name }}`;
-        }
-        // Can not use identifier here
-        return chSql`${clause.name} AS (${clause.sql})`;
-      }),
+      await Promise.all(
+        withClauses.map(async clause => {
+          // The sql logic can be specified as either a string, ChSql instance or a
+          // chart config object.
+          let resolvedSql: ChSql;
+          if (typeof clause.sql === 'string') {
+            resolvedSql = chSql`${{ Identifier: clause.sql }}`;
+          } else if (clause.sql && 'sql' in clause.sql) {
+            resolvedSql = clause.sql;
+          } else if (
+            clause.sql &&
+            ('select' in clause.sql || 'connection' in clause.sql)
```

Comment:
style: not a blocker. this isn't very readable imo. I wonder if we could split fields for `with` like
```
{
name: string;
sql?: string;
chSql?: ChSql;
chartConfig?: ChartConfig;
}
```